### PR TITLE
Reworked CMake infrastructure

### PR DESCRIPTION
**Pull request by pmccormick**
Here's a proposed set of changes to the CMake configuration.   It still needs work but is hopefully an improved path forward.  I have tested as broadly as I could (x86, aarch64, gcc, clang, CentOS, RHEL, MacOS -- waiting for a Power system to become available).  
 
I have not tested Windows environments, there is limited testing with shared libraries, non-make (e.g. ninja) builds need to be tested, and I most certainly haven't been through an exhaustive exploration of all the parameter spaces and nuances introduced by all LLVM configuration settings.  Testing (lit) needs to be incorporated.  AMLIR, FIR, etc.  that are part of the upstream plans all need to be added. 

My basic use cases have used a command line with these basic components: 

`$ cmake -DCMAKE_BUILD_TYPE=[Release|Debug|RelWithDebInfo] -DLLVM_DIR=/path/to/llvm/install-dir/lib/cmake/llvm -DCMAKE_INSTALL_PREFIX=/path/to/install-dir -DFLANG_INCLUDE_TESTS=ON -DCMAKE_VERBOSE_MAKEFILE=ON path/to/CMakeLists.txt`

### CODE REVIEW

**Comment by DavidTruby** [comment]
Hi Pat, this doesn't build for me in tree. I get errors saying that the flang headers cannot be found, e.g.: 
```
/home/davtru01/llvm-project/llvm/tools/f18/lib/Common/default-kinds.cpp:9:10: fatal error: 'flang/Common/default-kinds.h' file not found
#include "flang/Common/default-kinds.h"
```

Should these changes work for building this way?

**Comment by pmccormick** [comment]
Hi @DavidTruby.   Interesting, I didn't see this problem but I have a theory as to why -- a mistake on my side...  Let me poke at it some.  

Which branch of llvm-project are you using?  Thanks.

**Comment by DavidTruby** [comment]
This is with current llvm-project master, although I haven't pulled since yesterday afternoon. Thanks for taking a look!

**Comment by pmccormick** [comment]
@DavidTruby -- I reproduced your issue.  The following lines in the toplevel CMakeLists.txt file need to move outside the standalone build conditional: 

```
include_directories(BEFORE
  ${FLANG_BINARY_DIR}/include
  ${FLANG_SOURCE_DIR}/include) 
```
My mistake was a lurking install from a previous test build that let this slip past.  Whoops...

I'll push a fix after I look at one additional issue that popped up.

**Comment by schweitzpgi** [comment]
Thanks, Pat. I will merge these into my stream and run them through the wash on my end.

**Comment by tskeith** [comment]
When I run `ctest` I get "No test configuration file found!" and `ninja check-all` gets "error: unknown target 'check-all'".

**Comment by tskeith** [comment]
Can you update the build instructions in `README.md`? At least the part about `BUILD_WITH_CLANG` needs to be fixed.

**Comment by tskeith** [comment]
Building with clang I used to get these options that are no longer present. Is that intentional?
```
-fno-rtti
-fno-exceptions
-pedantic
-Wall
-Wextra
-Werror
-Wcast-qual
-Wimplicit-fallthrough
-Wdelete-non-virtual-dtor
```

Also, `-std=c++17` has changed to `-std=gnu++17`. What does that mean?

**Comment by DavidTruby** [comment]
> Also, -std=c++17 has changed to -std=gnu++17. What does that mean?

This means c++17 with gnu extensions, and is almost definitely not what we want. I think we need to set CMAKE_CXX_EXTENSIONS=OFF before setting cmake_cxx_standard?

**Comment by foxpm** [comment]
> > Also, -std=c++17 has changed to -std=gnu++17. What does that mean?
> 
> This means c++17 with gnu extensions, and is almost definitely not what we want. I think we need to set CMAKE_CXX_EXTENSIONS=OFF before setting cmake_cxx_standard?

Yes, most certainly.  I believe this is CMake's default behavior when using gcc and C++17 so we will certainly need to override it.  I'm swamped right now but will try to get these issues as soon as I can.

**Comment by DavidTruby** [comment]
Hi Pat, if you're swamped at the moment could I volunteer myself to have a look at this? We have an intention to merge on Monday now and we need to get this in first!

**Comment by pmccormick** [comment]
Thanks David -- absolutely!  That would be really helpful.  I'm home with family on a bit of a medical leave so I could use the help.   Let me know if you have any questions (Rich or Will can get you my direct contact info if you need it).  Thanks again!

**Comment by foxpm** [comment]
> After `ninja install`, all of the `install/include/flang/*/*.h` files are dangling symlinks.

@tskeith, I've been unable to reproduce this issue.  Can you provide some more specifics on your build configuration?  Thanks.

**Comment by tskeith** [comment]
> > After `ninja install`, all of the `install/include/flang/*/*.h` files are dangling symlinks.
> 
> @tskeith, I've been unable to reproduce this issue. Can you provide some more specifics on your build configuration? Thanks.

This is my cmake command:
```
cmake \
  -GNinja \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX=install \
  -DLLVM_DIR=$LLVM/lib/cmake/llvm \
  -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/9.0.0_1/bin/clang++ \
 ../..
```
Then after `ninja install`:
```
$ ls -lL install/include/flang/Common/Fortran-features.h
lrwxr-xr-x  1 tsk  staff  45 Mar  9 15:40 install/include/flang/Common/Fortran-features.h@ -> ../../include/flang/Common/Fortran-features.h
ls: Fortran-features.h: No such file or directory
```

**Comment by foxpm** [comment]
Still no luck for me...  I went so far as to install llvm@9 via homebrew to come as close as I could to your configuration details...  After `ninja install` the file, not a symlink, is there: 

>$ ls -lL install/include/flang/Common/Fortran-features.h
> -rw-r--r--  1 pat  staff  2976 Mar  3 19:55 install/include/flang/Common/Fortran-features.h
>$ 
>$ head -5 install/include/flang/Common/Fortran-features.h
>//===-- include/flang/Common/Fortran-features.h -----------------*- C++ -*-===//
>//
>// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
>// See https://llvm.org/LICENSE.txt for license information.
>// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

**Comment by DavidTruby** [comment]
Hi Tim, is this an in or out of tree build? I get the same results as Pat on my in-tree builds. These files shouldn't be symlinks at all, let alone dead links.

**Comment by tskeith** [comment]
> Still no luck for me... I went so far as to install llvm@9 via homebrew to come as close as I could to your configuration details... After `ninja install` the file, not a symlink, is there:

The problem was that I had some extraneous symlinks that were getting installed, I guess replacing the correct files.

So the dangling symlinks were just a symptom of a different problem: `.h` files are being install from `lib/*/*.h` and not just from `include/flang/*/*.h`. For example, `Semantics/resolve-names.h` is under `lib`, not `include/flang` in the git source. But in the install directory iends up in `include/flang/Semantics/resolve-names.h`.

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+  install(DIRECTORY common evaluate decimal parser semantics 
+    DESTINATION include/flang
+    COMPONENT flang-headers
+    FILES_MATCHING
+    PATTERN "*.def"
+    PATTERN "*.h"
+    PATTERN "config.h" EXCLUDE
+    PATTERN ".git"     EXCLUDE
+    PATTERN ".svn"     EXCLUDE)
+endif()
```
[the remark this answers is not in the document]
Comment (tskeith) [reply lib/CMakeLists.txt]: Right, this is installing `.h` files that shouldn't be installed.

**Comment by DavidTruby** [comment]
I don't get that file installed either, can you check with a clean build directory to see if that still happens?

**Comment by tskeith** [comment]
> I don't get that file installed either, can you check with a clean build directory to see if that still happens?

I deleted my build directory and built again and still get those .h files installed. This is an out-of-tree build.

I think the problem is in `lib/CMakeLists.txt` where I just commented (and Eric did previously).

**Comment by DavidTruby** [comment]
I see why. That install headers code shouldn't be there, but I still wasn't getting those headers installed as the directory names are lowecase there and I am building on Linux with a case sensitive filesystem. On OSX where it's case insensitive you see a different behaviour.

I'll remove those lines from that file.

**Comment by DavidTruby** [comment]
@tskeith The new commit I pushed just removes those lines from the lib directory, so only the include directory headers should get installed. 
(The force push was just a rebase on top of master)

**Comment by DavidTruby** [comment]
I've got a fix for llvm-lit tests not working, which I will push now. As regards the ctest tests, there are a lot of hardcoded paths in the .sh scripts that are not going to work in the new world. I'll change these .sh scripts to be .sh.in scripts configurable by cmake, but that'll take a little more work.

**Comment by DavidTruby** [comment]
Looks like I needed to add a check-all target when we're building standalone, so I've done that too.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,150 +1,332 @@
...
+  if (CXX_SUPPORTS_NO_NESTED_ANON_TYPES_FLAG)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-nested-anon-types")
+  endif()
+
+  # Add to or adjust build type flags. 
+  # 
+  # TODO: This needs some extra thought.  CMake's default for release builds 
+  # is -O3, which can cause build failures on certain platforms (and compilers)
+  # with the current code base -- some templated functions are inlined and don't
+  # become available at link time when using -O3 (with Clang under MacOS/darwin).
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: Those cases probably need looking at separately at some point: we really shouldn't be relying on inline functions being available at link time, as it's entirely valid for the compiler to not generate outline versions of them.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,150 +1,332 @@
...
+    COMPONENT flang-headers
+    FILES_MATCHING
+    PATTERN "*.def"
+    PATTERN "*.h"
+    PATTERN "*.inc"    
+    PATTERN "config.h" EXCLUDE
+    PATTERN ".git"     EXCLUDE
+    PATTERN "CMakeFiles" EXCLUDE    
+    PATTERN "*")
+endif()
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: I've removed this from lib/CMakeLists.txt

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,150 +1,332 @@
...
+  cmake_policy(SET CMP0075 NEW)
+endif()
+
+# option() honors normal variables.
+if (POLICY CMP0077) 
+  cmake_policy(SET CMP0077 NEW)
+endif()
+
+# Flang requires C++17. 
+set(CMAKE_CXX_STANDARD 17)
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: Are we sure that the linker errors were due to C++14 vs C++17? I don't think there are any ABI breaking changes between those two releases.

File: include/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+add_subdirectory(flang)
+# FIXME: upstream PRs should uncomment this...
+# add_subdirectory(optimizer)
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply include/CMakeLists.txt]: the include/flang subdirectory still has a lowercase letter (copying what clang and mlir do)

File: lib/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (DavidTruby) [reply lib/CMakeLists.txt]: I think I've fixed this now

**Comment by DavidTruby** [comment]
> Looks like I needed to add a check-all target when we're building standalone, so I've done that too.

This didn't work, but I think I have an idea for a fix

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply CMakeLists.txt]: 100%. `constexpr` is handled differently between the two language standards.

File: include/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply include/CMakeLists.txt]: optimizer is spelled wrong.

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: Is it handled differently? I know constexpr now implies the new inline variable semantics, but weren't those semantics explicitly added for constexpr variables in 14 anyway? (And just not available separately from constexpr)

[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: Regardless, I've tried with gcc 7, gcc 9, clang 8 and clang 9 and they all seem happy building LLVM with -std=c++14 and then building f18 with -std=c++17 and linking it all at the end. Perhaps the incompatibilities were fixed in the mean time on the LLVM side?

**Comment by DavidTruby** [comment]
@tskeith I believe the commits I have just pushed should fix testing for both in and out of tree builds. The process for running the tests is as follows:

# Out of tree
* ctest tests are run by running the ctest command in the top level of the build directory. There is no "test" target anymore.
* lit tests can be run by running the check-flang or check-all target. Either of these will only run the flang tests.

# In tree
* ctest tests can be run by running `cd tools/flang && ctest` from the top level build directory. There is no "test" target anymore.
* lit tests can be run by running the check-flang target for only the flang tests, or the check-all target for all the tests of enabled LLVM projects, including flang.

File: include/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (DavidTruby) [reply include/CMakeLists.txt]: I'll just remove this comment, as we should add this code in when we actually need it and not before

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,150 +1,332 @@
...
+    if (CXX_SUPPORTS_NO_STRING_CONVERSION_FLAG)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-string-conversion")
+    endif()
+  endif()  # Clang. 
+
+  check_cxx_compiler_flag("-Werror -Wnested-anon-types" CXX_SUPPORTS_NO_NESTED_ANON_TYPES_FLAG)
+  if (CXX_SUPPORTS_NO_NESTED_ANON_TYPES_FLAG)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-nested-anon-types")
+  endif()
+
```
[the remark this answers is not in the document]
Comment (DavidTruby) [reply CMakeLists.txt]: I think these are implied by the LLVM targets

**Comment by tskeith** [comment]
Has the question about compilation options been resolved?

Also, I'm getting this, building with clang 9.0:
```
warning: unknown warning option '-Wno-deprecated-copy'; did you mean '-Wno-deprecated'?
      [-Wunknown-warning-option]
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,154 +1,343 @@
...
+  # We need a pre-built/installed version of LLVM.
+  find_package(LLVM REQUIRED HINTS "${LLVM_CMAKE_PATH}")
+  list(APPEND CMAKE_MODULE_PATH ${LLVM_DIR})
+
+  # If LLVM links to zlib we need the imported targets so we can too.
+  if(LLVM_ENABLE_ZLIB)
+    find_package(ZLIB REQUIRED)
+  endif()
+
+  set(LLVM_DEFAULT_EXTERNAL_LIT ${LLVM_DIR}/../../../bin/llvm-lit)
```
Comment (tskeith) [inline CMakeLists.txt]: What is this for?

Comment (DavidTruby) [reply CMakeLists.txt]: No longer necessary and left in by mistake.

**Comment by tskeith** [comment]
`bin/flang` doesn't have the right `-intrinsic-module-directory` directory.

**Comment by DavidTruby** [comment]
> Has the question about compilation options been resolved?

Which question?
> 
> Also, I'm getting this, building with clang 9.0:
> 
> ```
> warning: unknown warning option '-Wno-deprecated-copy'; did you mean '-Wno-deprecated'?
>       [-Wunknown-warning-option]
> ```

I have also started seeing this and can't work it out for the life of me. I have this in the cmake file:
```
check_cxx_compiler_flag("-Werror -Wno-deprecated-copy" CXX_SUPPORTS_NO_DEPRECATED_COPY_FLAG)
if (CXX_SUPPORTS_NO_DEPRECATED_COPY_FLAG)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-copy")
endif()
```
And yet, this function reports success for clang 9, even though clang 9 obviously doesn't have this option after a simple test...

**Comment by tskeith** [comment]
> Building with clang I used to get these options that are no longer present. Is that intentional?
> 
> ```
> -fno-rtti
> -fno-exceptions
> -pedantic
> -Wall
> -Wextra
> -Werror
> -Wcast-qual
> -Wimplicit-fallthrough
> -Wdelete-non-virtual-dtor
> ```

**Comment by DavidTruby** [comment]
> > Building with clang I used to get these options that are no longer present. Is that intentional?
> > ```
> > -fno-rtti
> > -fno-exceptions
> > -pedantic
> > -Wall
> > -Wextra
> > -Werror
> > -Wcast-qual
> > -Wimplicit-fallthrough
> > -Wdelete-non-virtual-dtor
> > ```

sorry, I missed this comment somehow. I think we should still have those flags so I will re-add them

**Comment by DavidTruby** [comment]
On my system `cmake_cxx_source_compiles` (the function that `check_cxx_compiler_flag` calls) _always_ returns success. Even if I put completely bogus source in there...

**Comment by DavidTruby** [comment]
Looks like a cmake bug to me. I have a workaround that I have just pushed, can you check that the bad behaviour is gone for you too @tskeith ?

**Comment by tskeith** [comment]
I'm still seeing differences in compilation options.

**gcc 8.3.0**

These are duplicated:
  -D__STDC_CONSTANT_MACROS
  -D__STDC_FORMAT_MACROS
  -D__STDC_LIMIT_MACROS

These are added, along with a few others that don't look important:
  -fPIC
  -fno-semantic-interposition
  -fno-strict-aliasing

**clang 9.0.0**

These are duplicated:
  -D__STDC_CONSTANT_MACROS
  -D__STDC_FORMAT_MACROS
  -D__STDC_LIMIT_MACROS
  -Wall
  -Wcast-qual
  -Wcovered-switch-default
  -Wdelete-non-virtual-dtor
  -Wextra
  -Wimplicit-fallthrough
  -Wstring-conversion

This is added:
  -fPIC

The duplicated options probably aren't a big deal but they do have an impact when
using ccache. Adding or removing a duplicate option leads to unnecessary recompilations.
And it makes the compilation commands even harder to read than they already are.

**Comment by tskeith** [comment]
MinSizeRel builds don't seem to work at all.

**Comment by DavidTruby** [comment]
fPIC is the default on newer compilers but llvm adds it for older compilers. We're inheriting llvm's options so I can turn off where we manually add the ones llvm already has.

**Comment by DavidTruby** [comment]
I can't see why MinSizeRel shouldn't work as it should just set Os instead of O3. I can have a further look at that tomorrow

**Comment by tskeith** [comment]
This is what I'm seeing in a MinSizeRel build:

> FAILED: lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o
/usr/local/Cellar/llvm/9.0.0_1/bin/clang++  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I../../include -Iinclude -isystem /Users/tsk/llvm/f18-llvm3/build/Release-clang-9/install/include -fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor  -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -w -fdiagnostics-color -Wno-deprecated-copy -Wno-string-conversion -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -Os -DNDEBUG -DCHECK=(void) -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -std=c++17 -MD -MT lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o -MF lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o.d -o lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o -c ../../lib/Common/Fortran.cpp

> /bin/sh: -c: line 0: syntax error near unexpected token `('

> /bin/sh: -c: line 0: `/usr/local/Cellar/llvm/9.0.0_1/bin/clang++  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I../../include -Iinclude -isystem /Users/tsk/llvm/f18-llvm3/build/Release-clang-9/install/include -fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor  -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -w -fdiagnostics-color -Wno-deprecated-copy -Wno-string-conversion -Wno-unused-command-line-argument -Wstring-conversion           -Wcovered-switch-default -Wno-nested-anon-types -Os -DNDEBUG -DCHECK=(void) -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk   -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -std=c++17 -MD -MT lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o -MF lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o.d -o lib/Common/CMakeFiles/FortranCommon.dir/Fortran.cpp.o -c ../../lib/Common/Fortran.cpp'

**Comment by foxpm** [comment]
> This is what I'm seeing in a MinSizeRel build:
> ... 

I've duplicated this -- taking a look at it now...

**Comment by DavidTruby** [comment]
Looks like it's coming from the - DCHECK=(void), are we introducing that for the CHECK macro?

**Comment by DavidTruby** [comment]
Never mind, I'm too slow and Pat has already fixed it!

**Comment by DavidTruby** [comment]
I have rebased this with the FIR changes, and addressed review comments. 
If there are further issues which are non-breaking, can I suggest we note them and deal with them later so that we can get this in ready for the upstream merge?

**Comment by tskeith** [comment]
> I have rebased this with the FIR changes, and addressed review comments.
> If there are further issues which are non-breaking, can I suggest we note them and deal with them later so that we can get this in ready for the upstream merge?

What issues are still unresolved?

**Comment by DavidTruby** [comment]
> > I have rebased this with the FIR changes, and addressed review comments.
> > If there are further issues which are non-breaking, can I suggest we note them and deal with them later so that we can get this in ready for the upstream merge?
> 
> What issues are still unresolved?

Any that you still might have when reviewing what I've pushed now, plus the duplicated flags. 
I can't replicate the flags getting duplicated, and I don't think it's hugely bad that they are at least for now.

**Comment by tskeith** [comment]
What about build instructions?

Apart from that, I don't have any issues. But someone who knows something about cmake should approve this.

**Comment by DavidTruby** [comment]
> What about build instructions?

The build instructions in the Readme work fine for out of tree builds. In tree builds are a little more complex to set up because the cmake infrastructure isn't there on the LLVM side, so I suggest adding that readme change once it is (which will be at the same time as the merge)

**Comment by DavidTruby** [comment]
I'll add Eric's LLVM-side cmake changes as the first commit after the merge commit when I do the merge with the monorepo, and I'll push that all to llvm-project/master in one go.

**Comment by tskeith** [comment]
> The build instructions in the Readme work fine for out of tree builds.

The README says to set `CXX` or `BUILD_WITH_CLANG`. Does `BUILD_WITH_CLANG` still work?

**Comment by DavidTruby** [comment]
It doesn't, you should just set the CXX env var or CMAKE_CXX_COMPILER. I can change that in the readme.

**Comment by DavidTruby** [comment]
I think just removing the lines about -DBUILD_WITH_CLANG and -DGCC is sufficient to update the instructions.

**Comment by DavidTruby** [comment]
Is this ready for me to squash the commits?

**Comment by tskeith** [comment]
> I think just removing the lines about -DBUILD_WITH_CLANG and -DGCC is sufficient to update the instructions.

Now GCC is used in the cmake file but not mentioned in the instructions.

**Comment by DavidTruby** [comment]
I've removed the -DGCC flag. It won't work correctly with linking to LLVM anyway.

**Comment by tskeith** [comment]
These files are copied into the install directory. Are they supposed to be?
```
include/flang/CMakeLists.txt
include/flang/Optimizer/CMakeLists.txt
include/flang/Optimizer/.clang-format
include/flang/Optimizer/Dialect/CMakeLists.txt
include/flang/Optimizer/Dialect/FIROps.td
include/flang/Config/config.h.cmake
include/flang/Version.inc.in
include/flang/Lower/.clang-format
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -102,62 +144,227 @@ if(LINK_WITH_FIR)
...
+configure_file(
+  ${CMAKE_CURRENT_SOURCE_DIR}/include/flang/Version.inc.in
+  ${CMAKE_CURRENT_BINARY_DIR}/include/flang/Version.inc)
+# Configure Flang's version info header file.
 configure_file(
   ${FLANG_SOURCE_DIR}/include/flang/Config/config.h.cmake
   ${FLANG_BINARY_DIR}/include/flang/Config/config.h)
+
+# Add global F18 flags.
+set(CMAKE_CXX_FLAGS "-fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor ${CMAKE_CXX_FLAGS}")
```
Comment (isuruf) [inline CMakeLists.txt]: Is it possible to add checks for these flags before using them?
Also, is `-fno-rtti` still needed? `functional` header cannot be included in Windows with this flag.

Comment (DavidTruby) [reply CMakeLists.txt]: I'll need to look in to what LLVM does here. We do need -fno-rtti as llvm builds with that, at least on Linux, macOS and *BSD. I don't suppose you could send a list of what flags LLVM uses by default on Windows?

Comment (schweitzpgi) [reply CMakeLists.txt]: I think it might be. My build environment was tripping over some RTTI link failures without it. YMMV.

**Comment by DavidTruby** [comment]
> These files are copied into the install directory. Are they supposed to be?
> 
> ```
> include/flang/CMakeLists.txt
> include/flang/Optimizer/CMakeLists.txt
> include/flang/Optimizer/.clang-format
> include/flang/Optimizer/Dialect/CMakeLists.txt
> include/flang/Optimizer/Dialect/FIROps.td
> include/flang/Config/config.h.cmake
> include/flang/Version.inc.in
> include/flang/Lower/.clang-format
> ```

This should be fixed now. The .td file still gets installed, as that is what LLVM, Clang and MLIR do with their .td files. The rest don't get installed anymore.

File: CMakeLists.txt (hunk 8 above)
Comment (isuruf) [reply CMakeLists.txt]: I'm cross compiling at the moment, so lots of flags are added from my build environment.
LLVM's RTTI handling is at https://github.com/llvm/llvm-project/blob/50a6d3a6486d81d21a2c31ce8522321e19bed35e/llvm/cmake/modules/AddLLVM.cmake#L38-L53. Exception handling https://github.com/llvm/llvm-project/blob/50a6d3a6486d81d21a2c31ce8522321e19bed35e/llvm/cmake/modules/AddLLVM.cmake#L13-L36

Comment (DavidTruby) [reply CMakeLists.txt]: Would it be OK with you if we merged this now and opened another issue/PR to fix windows build issues? I'll find out if I can set up a native Windows build environment at some point to debug these issues.

Comment (isuruf) [reply CMakeLists.txt]: Sure

**Comment by DavidTruby** [comment]
Squashed and rebase, ready to merge

File: tools/f18/flang.sh.in [hunk 9]
```diff
@@ -26,4 +26,4 @@ function abspath() {
 
 wd=`abspath $(dirname "$0")/..`
 
-${wd}/bin/f18 -module-suffix .f18.mod -intrinsic-module-directory ${wd}/include $*
+${wd}/bin/f18 -module-suffix .f18.mod -intrinsic-module-directory @FLANG_INTRINSIC_MODULES_DIR@ $*
```
Comment (tskeith) [inline tools/f18/flang.sh.in]: Why is the intrinsic module directory a cmake variable substitution but the path to f18 still relative to this script? Shouldn't they both be found using a similar mechanism?

Comment (DavidTruby) [reply tools/f18/flang.sh.in]: The f18 binary will always be in build_dir/bin/f18, whereas the location of the modules is different if you're in an in-tree or out-of-tree build. Strictly speaking you're probably correct, but the f18 path will be correct for both builds here so it doesn't really matter.

**Comment by sscalpone** [comment]
@pmccormick @DavidTruby What is the change with %S and %B in the test files?

**Comment by sscalpone** [comment]
@pmccormick Is the commit message the way you want it?  I don't have a suggestion, I'm just asking.

**Comment by DavidTruby** [comment]
> What is the change with %S and %B in the test files?

The script files sometimes need certain paths correcting (hence the .in files), and if the configured versions of these get output to the source directory it messes up if you have multiple build directories. As a result I had them get output to the build directory, but then the tests need the path to the build directory so they can find them. So I set %B to the build directory, to match the %S that always exists in lit and points to the source directory.

**Comment by DavidTruby** [comment]
@sscalpone rebased on recent changes

**Comment by tskeith** [comment]
When I build out-of-tree and run the lit tests I get 207 failures with errors like this:

> /Users/tsk/f18/pmccormick-master/build/Release/test/Semantics/Output/allocate01.f90.script: line 1: /Users/tsk/f18/pmccormick-master/build/Release/test/Semantics/test_errors.sh: No such file or directory

Am I doing something wrong?

**Comment by tskeith** [comment]
> > What is the change with %S and %B in the test files?
> 
> The script files sometimes need certain paths correcting (hence the .in files), and if the configured versions of these get output to the source directory it messes up if you have multiple build directories. As a result I had them get output to the build directory, but then the tests need the path to the build directory so they can find them. So I set %B to the build directory, to match the %S that always exists in lit and points to the source directory.

This should be explained in the commit message.

**Comment by DavidTruby** [comment]
Looks like the lit merge removed my CMakeLists.txt file that generates that file. Should be fixed now.

**Comment by sscalpone** [comment]
@pmccormick @DavidTruby How are you testing this change?  What is expected to work and not work?

**Comment by pmccormick** [comment]
@sscalpone, since we don't yet have a local buildbot-like system, I have been testing by hand on multiple platforms that I have easy access to (x86, arm) and across what OSes they support (darwin, ubuntu 18.04, CentOS 7, RHEL).  This includes standalone and mono-repo configurations.  Most things I have tried seem to work OK but I have not yet fully explored shared libraries.  The key question is what other build options are you looking for/expecting?

**Comment by DavidTruby** [comment]
I've tested on Linux and macOS, on x86 and Arm, with gcc 7,8,9 and clang 7,8,9 with in and out of tree builds. All these configurations should work.
I have also not explored shared libraries as they weren't working with MLIR last time I tried.

**Comment by sscalpone** [comment]
I'm thinking of the various CMAKE variable settings that LLVM supports, in tree and out of tree, using the build area or install area as LLVM_DIR (or whatever is current), setting flang and mlir as project to build, does it require clang or any others, ....   And I'll check for any variables that are used in the flang cmake files.

**Comment by DavidTruby** [comment]
It does require mlir to be built, but not clang. I've tested both with clang enabled and clang not enabled. We will as developers want to enable clang to get clang-format and git-clang-format though. 
In terms of the wide number of CMake settings that LLVM has... I think that's pretty much impossible to test until we get in-tree and can convince people to turn flang on in their buildbots and see what happens. I've tested it with most of the common ones (except LLVM_ENABLE_SHARED) including the 3 supported LLVM_USE_LINKER flags, but there's just too wide a vector to test every combination.

**Comment by sscalpone** [comment]
@pmccormick @DavidTruby Works for me on mac.  I'm ready to merge after these tests are updated:

    Flang :: Semantics/entry01.f90
    Flang :: Semantics/resolve78.f90
    Flang :: Semantics/resolve79.f90
    Flang :: Semantics/resolve80.f90
    Flang :: Semantics/resolve81.f90
    Flang :: Semantics/resolve82.f90
    Flang :: Semantics/resolve83.f90
    Flang :: Semantics/resolve84.f90
    Flang :: Semantics/resolve85.f90

**Comment by DavidTruby** [comment]
@sscalpone thanks for pointing those out, done!

**Pull request merged**